From e4e4b168708162ee87c5d3647569b283babe4740 Mon Sep 17 00:00:00 2001
From: tm <tm@x>
Date: Tue, 6 Oct 2026 17:52:30 +0100
Subject: [PATCH 16/20] Special forms are not shadowed by local variables at
 the head of a form (the expansions of the macros use them); boot image
 regenerated

---
 compiler.lsp |  9 +++++++--
 flisp.boot   | 11 ++++++-----
 system.lsp   |  7 ++++++-
 3 files changed, 19 insertions(+), 8 deletions(-)

diff --git a/compiler.lsp b/compiler.lsp
index 375ee13..63b2b7b 100644
--- a/compiler.lsp
+++ b/compiler.lsp
@@ -467,6 +467,9 @@
 	      (let ((c (aref v (- (length v) 1))))
 		(and (pair? c) (eq? (car c) '%source) (cdr c)))))))
 
+; TeXmacs: a local variable does not shadow a special form at the head of a
+; form: the macros expand to special forms (cond to begin...), which a local
+; variable named begin would capture otherwise
 (define special-forms
   '(quote if begin prog1 lambda and or while for return set! define trycatch))
 
@@ -487,8 +490,10 @@
 	       ((eof-object? x)
 		(compile-in g env tail? (list (top-level-value 'eof-object))))
 	       (else        (emit g 'loadv x))))
-	((or (not (symbol? (car x))) (bound? (resolve-global (car x)))
-	     (in-env? (car x) env))
+	((or (not (symbol? (car x)))
+	     (and (not (memq (car x) special-forms))
+		  (or (bound? (resolve-global (car x)))
+		      (in-env? (car x) env))))
 	 (compile-app g env tail? x))
 	; TeXmacs: a call of a function which is not defined yet (or of a
 	; macro defined later): the hook may give another form to compile
diff --git a/flisp.boot b/flisp.boot
index c4a92a3..84c6504 100644
--- a/flisp.boot
+++ b/flisp.boot
@@ -114,10 +114,10 @@
 	%fl:compile-if #68=#fn("<000r4c0qe1|31e1|31e2g331e3g331e4g331F6;0e5g331560e63045;" #(#fn(";000r5g2]‚>0e0~\x7fi02g344;g2^‚>0e0~\x7fi02g444;e0~\x7f^g2342e1~c2|332e0~\x7fi02g3342i026<0e1~c3325:0e1~c4}332e5~|322e0~\x7fi02g4342e5~}42;" #(%fl:compile-in
   %fl:emit brf ret jmp %fl:mark-label)) %fl:make-label %fl:cadr %fl:caddr
   %fl:cdddr %fl:cadddr %fl:void) compile-if)
-	%fl:compile-in #69=#fn(";000r4g3C6=0e0|}g3c144;g3?6¯0g3`‚:0e2|c342;g3a‚:0e2|c442;g3]‚:0e2|c542;g3^‚:0e2|c642;g3_‚:0e2|c742;e8g3316<0e2|c9g343;c:g3316C0e;|}g2c<c=31L144;e2|c>g343;g3MC@17H02e?g3M31E17;02e@g3M}326=0eA|}g2g344;cBqcCg3MeD32@16:02eEg3}3241;" #(%fl:compile-sym
+	%fl:compile-in #69=#fn(";000r4g3C6=0e0|}g3c144;g3?6¯0g3`‚:0e2|c342;g3a‚:0e2|c442;g3]‚:0e2|c542;g3^‚:0e2|c642;g3_‚:0e2|c742;e8g3316<0e2|c9g343;c:g3316C0e;|}g2c<c=31L144;e2|c>g343;g3MC@17W02c?g3Me@32@16H02eAg3M31E17;02eBg3M}326=0eC|}g2g344;cDqc?g3Me@32@16:02eEg3}3241;" #(%fl:compile-sym
   #(loada loadc loadg) %fl:emit load0 load1 loadt loadf loadnil %fl:fits-i8
   loadi8 #fn(eof-object?) %fl:compile-in #fn(top-level-value) eof-object loadv
-  resolve-global %fl:in-env? %fl:compile-app #fn(":000r1|6=0e0~\x7fi02|44;c1qi03M41;" #(%fl:compile-in
+  #fn(memq) special-forms resolve-global %fl:in-env? %fl:compile-app #fn(":000r1|6=0e0~\x7fi02|44;c1qi03M41;" #(%fl:compile-in
   #fn("=000r1|c0‚f0e1e2i1331316G0e3i10i11i12e2i133144;e4i10c5e2i133143;|c6‚C0e7i10i11i12i1344;|c8‚D0e9i10i11i12i13N44;|c:‚@0e;i10i11i1343;|c<‚=0e=c>qc?q42;|c@‚D0eAi10i11i12i13N44;|cB‚D0eCi10i11i12i13N44;|cD‚N0eEi10i11e2i1331c8eFi1331K44;|cG‚R0eHi10i11e2i1331eIi1331eJi133145;|cK‚O0e3i10i11]e2i1331342e4i10cL42;|cM‚m0e3i10i11^eIi1331342e2i1331C17902eNcO312ePi10i11e2i1331cQ44;|cR‚G0e3i10i11i12eSi133144;|cT‚€0e3i10i11^c<_e2i1331L3342eUeIi133131660^580eNcV312e3i10i11^eIi1331342e4i10cT42;eWi10i11i12i1344;" #(quote
   %fl:self-evaluating? %fl:cadr %fl:compile-in %fl:emit loadv if %fl:compile-if
   begin %fl:compile-begin prog1 %fl:compile-prog1 lambda %fl:call-with-values
@@ -129,7 +129,7 @@
   %fl:compile-for %fl:caddr %fl:cadddr return ret set! %fl:error "set!: second argument must be a symbol"
   %fl:compile-sym #(seta setc setg) define %fl:expand-define trycatch
   %fl:1arg-lambda? "trycatch: second form must be a 1-argument lambda"
-  %fl:compile-app)))) #fn(memq) special-forms compile-unknown-call) compile-in)
+  %fl:compile-app)))) compile-unknown-call) compile-in)
 	%fl:compile-or #70=#fn("<000r4e0|}g2g3^c146;" #(%fl:compile-short-circuit
 							brt) compile-or)
 	%fl:compile-prog1 #71=#fn(";000r3e0|}^e1g231342e2g231F6H0e3|}^e2g231342e4|c542;];" #(%fl:compile-in
@@ -224,8 +224,9 @@
   %fl:cddr #fn(nconc) #fn(map) #fn(":000r1|Me0i2:e1|31i11323130i11L3;" #(%fl:compile-thunk
   %fl:cadr)))) %fl:cadr) expand-let-syntax)
   #fn("6000r2|;" #() local-expansion-env)
-  #fn("7000r2|?640|;c0q|M41;" #(#fn("9000r1c0qc1|\x7f3241;" #(#fn("7000r1c0qc1q41;" #(#fn(":000r1~16602~NF6P0i3:e0~31i20NQ2i39e1~31i213242;~17E02i10C@17;02e2i1031E660|40;c3qe4i203141;" #(%fl:cadr
-  %fl:caddr resolve-global #fn("9000r1|6C0i4:e0|i3032i3142;i20c1‚60i30;i20c2‚>0i46i30i3142;i20c3‚>0i47i30i3142;i20c4‚>0i48i30i3142;~40;" #(%fl:expand-macro-call
+  #fn("7000r2|?640|;c0q|M41;" #(#fn("9000r1c0qc1|\x7f3241;" #(#fn("7000r1c0qc1q41;" #(#fn(":000r1c0i10c1326;0c2qi1041;~16602~NF6P0i3:e3~31i20NQ2i39e4~31i213242;~17E02i10C@17;02e5i1031E660|40;c6qe7i203141;" #(#fn(memq)
+  (quote lambda define) #fn("8000r1|c0=660i30;|c1=6>0i46i30i3142;i47i30i3142;" #(quote
+  lambda)) %fl:cadr %fl:caddr resolve-global #fn("9000r1|6C0i4:e0|i3032i3142;i20c1‚60i30;i20c2‚>0i46i30i3142;i20c3‚>0i47i30i3142;i20c4‚>0i48i30i3142;~40;" #(%fl:expand-macro-call
   quote lambda define let-syntax)) %fl:macrocall?))
   #fn("7000r0c0q]31i2041;" #(#fn("6000r1c0qm02|;" #(#fn("9000r1|?640|;|M?670|M5<0i4:|Mi3132~|N31K;" #())))))))
 							      #fn(assq)))) expand-in)))) expand)
diff --git a/system.lsp b/system.lsp
index 53facab..e24addf 100644
--- a/system.lsp
+++ b/system.lsp
@@ -907,7 +907,12 @@
 					  (car e)
 					  (expand-in (car e) env))
 				      (loop (cdr e))))))))
-	  (cond ((and bnd (pair? (cdr bnd)))  ; local macro
+	  (cond ((memq head '(quote lambda define))  ; never shadowed (TeXmacs)
+		 (case head
+		   ((quote)  e)
+		   ((lambda) (expand-lambda e env))
+		   (else     (expand-define e env))))
+		((and bnd (pair? (cdr bnd)))  ; local macro
 		 (expand-in (apply (cadr bnd) (cdr e))
 			    (local-expansion-env (caddr bnd) env)))
 		((or bnd                      ; bound lexical or toplevel var
-- 
2.50.1 (Apple Git-155)

